src/main/java/net/fabricmc/mappingio/adapter/EmptyEntryFilter.java

```diff
@@ -0,0 +1,280 @@
+/*
+ * Copyright (c) 2025 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.mappingio.adapter;
+
+import java.io.IOException;
+import java.util.EnumSet;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import org.jetbrains.annotations.Nullable;
+
+import net.fabricmc.mappingio.MappedElementKind;
+import net.fabricmc.mappingio.MappingFlag;
+import net.fabricmc.mappingio.MappingVisitor;
+
+/**
+ * A mapping visitor that filters out elements which effectively don't contain any data.
+ *
+ * @implNote This visitor requires one pre-pass in which it determines which elements contain data worth forwarding.
+ */
+public final class EmptyEntryFilter extends ForwardingMappingVisitor {
+	/**
+	 * Creates a new {@link EmptyEntryFilter} that treats destination names and descriptors which are equal to their source counterparts as empty.
+	 *
+	 * @param next The next visitor to forward the data to.
+	 */
+	public EmptyEntryFilter(MappingVisitor next) {
+		this(next, true);
+	}
+
+	/**
+	 * @param next The next visitor to forward the data to.
+	 * @param treatSrcOnDstAsEmpty Whether destination names and descriptors that are equal to their source counterparts should be treated as empty.
+	 */
+	public EmptyEntryFilter(MappingVisitor next, boolean treatSrcOnDstAsEmpty) {
+		super(next);
+		this.treatSrcOnDstAsEmpty = treatSrcOnDstAsEmpty;
+	}
+
+	@Override
+	public Set<MappingFlag> getFlags() {
+		Set<MappingFlag> ret = EnumSet.noneOf(MappingFlag.class);
+		ret.addAll(next.getFlags());
+		ret.add(MappingFlag.NEEDS_MULTIPLE_PASSES);
+
+		return ret;
+	}
+
+	@Override
+	public boolean visitHeader() throws IOException {
+		clsCounter = memberCounter = localCounter = -129; // lowest cached Integer by HotSpot - 1
```

Comment:
Clever 🤔

Reply:
Probably a pointless micro-optimization, but it doesn't hurt either